infoschema/builder.go

```diff
@@ -187,31 +189,31 @@ func (b *Builder) applyCreateTable(m *meta.Meta, roDBInfo *model.DBInfo, tableID
 
 	newTbl, ok := b.is.TableByID(tableID)
 	if ok {
-		roDBInfo.Tables = append(roDBInfo.Tables, newTbl.Meta())
+		dbInfo.Tables = append(dbInfo.Tables, newTbl.Meta())
 	}
 	return nil
 }
 
-func (b *Builder) applyDropTable(roDBInfo *model.DBInfo, tableID int64) {
+func (b *Builder) applyDropTable(dbInfo *model.DBInfo, tableID int64) {
 	bucketIdx := tableBucketIdx(tableID)
 	sortedTbls := b.is.sortedTablesBuckets[bucketIdx]
 	idx := sortedTbls.searchTable(tableID)
 	if idx == -1 {
 		return
 	}
-	if tableNames, ok := b.is.schemaMap[roDBInfo.Name.L]; ok {
+	if tableNames, ok := b.is.schemaMap[dbInfo.Name.L]; ok {
 		delete(tableNames.tables, sortedTbls[idx].Meta().Name.L)
 	}
 	// Remove the table in sorted table slice.
 	b.is.sortedTablesBuckets[bucketIdx] = append(sortedTbls[0:idx], sortedTbls[idx+1:]...)
 
 	// The old DBInfo still holds a reference to old table info, we need to remove it.
-	for i, tblInfo := range roDBInfo.Tables {
+	for i, tblInfo := range dbInfo.Tables {
 		if tblInfo.ID == tableID {
-			if i == len(roDBInfo.Tables)-1 {
-				roDBInfo.Tables = roDBInfo.Tables[:i]
+			if i == len(dbInfo.Tables)-1 {
+				dbInfo.Tables = dbInfo.Tables[:i]
 			} else {
-				roDBInfo.Tables = append(roDBInfo.Tables[:i], roDBInfo.Tables[i+1:]...)
+				dbInfo.Tables = append(dbInfo.Tables[:i], dbInfo.Tables[i+1:]...)
```

Comment:
Or here?
In the PR description, you'd better post the data race log, so it's easier for us to understand where is the cause of the problem.

Reply:
It actually happens here in the log, but I browse the file and find more places that may have the race.